src/collectors/github_runners.rs

```diff
@@ -0,0 +1,207 @@
+            loop {
+                if let Err(e) = rv2.update_stats().await {
+                    error!("{:#?}", e);
+                }
+
+                tokio::time::delay_for(Duration::from_secs(refresh_rate)).await;
+            }
+        });
+
+        Ok(rv)
+    }
+
+    async fn update_stats(&mut self) -> Result<()> {
+        let mut gauges = Vec::with_capacity(self.repos.len() * 2);
+        for repo in self.repos.iter() {
+            let mut url: Option<String> = String::from(GH_RUNNERS_ENDPOINT)
+                .replace("{owner_repo}", repo)
+                .into();
+
+            debug!("Updating runner's stats");
+
+            while url.is_some() {
+                let response = self
+                    .http
+                    .get(&url.unwrap())
+                    .headers(default_headers(&self.token))
+                    .send()
+                    .await?;
+
+                guard_rate_limited(&response)?;
+
+                url = next_uri(response.headers().get(LINK));
+                let resp = response.json::<ApiResponse>().await?;
+
+                for runner in resp.runners.iter() {
+                    let online = metric_factory(
+                        "online",
+                        "runner is online",
+                        &self.desc.fq_name,
+                        &repo,
+                        &runner.name,
+                    );
+                    online.set(if runner.status == "online" { 1 } else { 0 });
+                    gauges.push(online);
+
+                    let busy = metric_factory(
+                        "busy",
+                        "runner is busy",
+                        &self.desc.fq_name,
+                        &repo,
+                        &runner.name,
+                    );
+                    busy.set(if runner.busy { 1 } else { 0 });
+                    gauges.push(busy);
+                }
+            }
+        }
+
+        // lock and replace old data
+        let mut guard = self.metrics.write().unwrap();
+        *guard = gauges;
+
+        Ok(())
+    }
+}
+
+impl Collector for GithubRunners {
+    fn desc(&self) -> Vec<&Desc> {
+        vec![&self.desc]
+    }
+
+    fn collect(&self) -> Vec<MetricFamily> {
+        self.metrics.read().map_or_else(
+            |e| {
+                error!("Unable to collect: {:#?}", e);
+                Vec::with_capacity(0)
+            },
+            |guard| {
+                guard.iter().fold(Vec::new(), |mut acc, item| {
+                    acc.extend(item.collect());
+                    acc
+                })
+            },
+        )
+    }
+}
+
+fn guard_rate_limited(response: &Response) -> Result<()> {
+    if response.status() == StatusCode::FORBIDDEN {
+        warn!("403 FORBIDDEN response status");
+
+        let rate_limited = match response.headers().get("x-ratelimit-remaining") {
+            Some(rl) => rl.to_str()?.parse::<usize>()? == 0,
+            None => unreachable!(),
```

Comment:
Hmm, we could also get a forbidden if we're not rate limited, for example if the token is revoked. If the rate limit is not zero I'd put `false` here and raise an error with `response.error_for_status_ref()?;` above.

Reply:
you totally right, I've removed the if, and this function only checks if we are rate limited, if so we return request error with context that we limited